### PR TITLE
✨ feat: migrate less to token for Transfer

### PR DESCRIPTION
**Pull request by heiyu4585**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 💡 Background and solution
https://github.com/ant-design/ant-design/issues/41884

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     migrate less to token for Transfer      |
| 🇨🇳 Chinese |     migrate less to token for Transfer      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at d1f4134</samp>

This pull request adds a new demo and documentation for customizing the theme tokens of the `Transfer` component. It also refactors the style file of the component to use a consistent and concise naming convention and to simplify the style and token objects.

### <samp>🤖 Generated by Copilot at d1f4134</samp>

*  Add a new component token demo for the Transfer component, which shows how to customize the theme tokens for the component ([link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-05763f96de9838a4304bb2721b5ac0408faf555a5075f1497936c771a9370a68R1-R7), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-edca56351272e05c831757bdcbfe5641e3d7615af8bfaa391b4ad9ef8b9eb49cR1-R59), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-c165dbeb24339c042fcc975a9b661af8947ea942457cf8935f29817ac86033a9R35))
*  Refactor the style file for the Transfer component to use a more consistent and concise naming convention for the theme tokens, and to add some new properties for customizing the appearance of the component ([link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL3-R5), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL11-R23), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL92-R101), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL133-R141), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL194-R199), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL224-R228), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL283-R287), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL329-R333), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL351-R360), [link](https://github.com/ant-design/ant-design/pull/42431/files?diff=unified&w=0#diff-d9d90f3711c4b10ecac467dc5d915fd5e15d8398916ed340a5ca70968394f2cbL368-R381))

### CODE REVIEW

**Comment by MadCcc** [comment]
迁移文档需要补一下

**Comment by heiyu4585** [comment]
> 迁移文档需要补一下

done

File: docs/react/migrate-less-variables.en-US.md [hunk 1]
```diff
@@ -329,7 +329,19 @@ This document contains the correspondence between all the less variables related
...
+| Less variables | Component Token | Note |
+| --- | --- | --- |
+| `@transfer-header-height` | `transferheaderHeight` | - |
+| `@transfer-item-height` | `transferItemHeight` | Global Token |
+| `@transfer-disabled-bg` | - | Deprecated |
+| `@transfer-list-height` | - | Deprecated |
+| `@transfer-item-hover-bg` | - | Deprecated |
+| `@transfer-item-selected-hover-bg` | - | Deprecated |
+| `@transfer-item-padding-vertical` | `transferItemVerticalPadding` | - |
+| `@transfer-list-search-icon-top` | - | Deprecated |
```
[the remark this answers is not in the document]
Comment (MadCcc) [reply docs/react/migrate-less-variables.en-US.md:344]: 剩下的应该都可以迁移，从 v4 里找一下用法加回来吧

[the remark this answers is not in the document]
Comment (heiyu4585) [reply docs/react/migrate-less-variables.en-US.md]: done

File: docs/react/migrate-less-variables.zh-CN.md [hunk 2]
```diff
@@ -327,7 +327,19 @@ title: Less 变量迁移 Design Token
...
+<!-- prettier-ignore -->
+| Less variables | Component Token | Note |
+| --- | --- | --- |
+| `@transfer-header-height` | `headerHeight` | - |
+| `@transfer-item-height` | `itemHeight` | - |
+| `@transfer-disabled-bg` | `disabledBg` | - |
+| `@transfer-list-height` | `listHeight` | - |
+| `@transfer-item-hover-bg` | `itemHoverBg` | - |
+| `@transfer-item-selected-hover-bg` | `itemSelectedHoverBg` | - |
+| `@transfer-item-padding-vertical` | `itemVerticalPadding` | - |
```
Comment (MadCcc) [inline docs/react/migrate-less-variables.zh-CN.md:341]:
```suggestion
| `@transfer-item-padding-vertical` | `itemPaddingBlock` | - |
```
这个记得改一下

File: components/transfer/style/index.ts [hunk 3]
```diff
@@ -1,20 +1,24 @@
...
-import { resetComponent, resetIcon, textEllipsis } from '../../style';
 
 export interface ComponentToken {
   listWidth: number;
   listWidthLG: number;
   listHeight: number;
+  itemHeight: number;
+  itemPaddingBlock: number;
+  headerHeight: number;
+  itemPaddingVertical: number;
```
Comment (MadCcc) [inline components/transfer/style/index.ts:13]:
```suggestion
```
和 itemPaddingBlock 重复了

File: components/transfer/style/index.ts [hunk 4]
```diff
@@ -1,20 +1,24 @@
...
+  disabledBg: string;
```
Comment (MadCcc) [inline components/transfer/style/index.ts:14]: 这个没有用过

File: components/transfer/style/index.ts [hunk 5]
```diff
@@ -1,20 +1,24 @@
...
+  itemHoverBg: string;
```
Comment (MadCcc) [inline components/transfer/style/index.ts:15]: 这个也没用过

File: components/transfer/style/index.ts [hunk 6]
```diff
@@ -1,20 +1,24 @@
...
+  itemSelectedHoverBg: string;
+  listSearchIconTop: number;
```
Comment (MadCcc) [inline components/transfer/style/index.ts:17]: 同上

File: components/transfer/style/index.ts (hunk 3 above)
Comment (heiyu4585) [reply components/transfer/style/index.ts]: done

**Comment by heiyu4585** [comment]
以上已完成

File: components/transfer/style/index.ts [hunk 7]
```diff
@@ -250,7 +253,7 @@ const genTransferListStyle: GenerateStyle<TransferToken> = (token: TransferToken
       // Do not change hover style when `oneWay` mode
       [`&-show-remove ${componentCls}-list-content-item:not(${componentCls}-list-content-item-disabled):hover`]:
         {
-          background: 'transparent',
+          background: itemHoverBg,
```
Comment (MadCcc) [inline components/transfer/style/index.ts:256]: 这里不一样了吧，原本是透明的

File: components/transfer/style/index.ts [hunk 8]
```diff
@@ -221,7 +224,7 @@ const genTransferListStyle: GenerateStyle<TransferToken> = (token: TransferToken
 
           '&::after': {
             position: 'absolute',
-            insert: `-${transferItemPaddingVertical}px -50%`,
+            insert: `-${itemPaddingBlock}px -50%`,
```
Comment (MadCcc) [inline components/transfer/style/index.ts:227]:
```suggestion
            inset: `-${itemPaddingBlock}px -50%`,
```
估计一开始就写错了。。

File: components/transfer/style/index.ts [hunk 9]
```diff
@@ -365,9 +362,26 @@ export default genComponentStyleHook(
...
+    } = token;
+    const fontHeight = Math.round(fontSize * lineHeight);
+    return {
+      listWidth: 180,
+      listHeight: 200,
+      listWidthLG: 250,
+      headerHeight: controlHeightLG,
+      itemHeight: controlHeight,
+      itemPaddingBlock: (controlHeight - fontHeight) / 2,
+      disabledBg: colorBgContainerDisabled,
```
Comment (MadCcc) [inline components/transfer/style/index.ts:382]: 这个就直接用 colorBgContainerDisabled 这个全局变量吧，不添加 component token 了

Comment (heiyu4585) [reply components/transfer/style/index.ts]: done

File: components/transfer/style/index.ts [hunk 10]
```diff
@@ -221,14 +223,14 @@ const genTransferListStyle: GenerateStyle<TransferToken> = (token: TransferToken
...
-            insert: `-${transferItemPaddingVertical}px -50%`,
+            inset: `-${itemPaddingBlock}px -50%`,
             content: '""',
           },
         },
 
         [`&:not(${componentCls}-list-content-item-disabled)`]: {
           '&:hover': {
-            backgroundColor: token.controlItemBgHover,
+            backgroundColor: colorBgContainerDisabled,
```
Comment (MadCcc) [inline components/transfer/style/index.ts:233]: 这个不应该是 disabled 的样式，应该还是 `controlItemBgHover`

File: components/transfer/style/index.ts [hunk 11]
```diff
@@ -1,20 +1,21 @@
...
 
 export interface ComponentToken {
   listWidth: number;
   listWidthLG: number;
   listHeight: number;
+  itemHeight: number;
+  itemPaddingBlock: number;
+  headerHeight: number;
+  itemHoverBg: string;
+  itemSelectedHoverBg: string;
```
Comment (MadCcc) [inline components/transfer/style/index.ts:14]: 这两个 token 还没有在 style 里用过

Comment (heiyu4585) [reply components/transfer/style/index.ts:14]: done

File: components/transfer/style/index.ts (hunk 10 above)
Comment (heiyu4585) [reply components/transfer/style/index.ts]: done

File: components/transfer/style/index.ts [hunk 12]
```diff
@@ -365,9 +364,18 @@ export default genComponentStyleHook(
...
+    const fontHeight = Math.round(fontSize * lineHeight);
+    return {
+      listWidth: 180,
+      listHeight: 200,
+      listWidthLG: 250,
+      headerHeight: controlHeightLG,
+      itemHeight: controlHeight,
+      itemPaddingBlock: (controlHeight - fontHeight) / 2,
+      itemHoverBg: controlItemBgHover,
+      itemSelectedHoverBg: controlItemBgHover,
```
Comment (MadCcc) [inline components/transfer/style/index.ts:378]: 这里写错了，是 `controlItemBgActiveHover`

Comment (heiyu4585) [reply components/transfer/style/index.ts]: done

**Review by MadCcc — APPROVED** [review]

**Pull request merged**